files.go

```diff
@@ -0,0 +1,99 @@
+package main
+
+import (
+	"fmt"
+	"io/ioutil"
+	"net/url"
+	"path/filepath"
+	"strings"
+
+	"github.com/Azure/custom-script-extension-linux/blobutil"
+	"github.com/Azure/custom-script-extension-linux/download"
+	"github.com/Azure/custom-script-extension-linux/preprocess"
+	"github.com/pkg/errors"
+)
+
+// downloadAndProcessURL downloads using the specified downloader and saves it to the
+// specified existing directory, which must be the path to the saved file. Then
+// it post-processes file based on heuristics.
+func downloadAndProcessURL(url, downloadDir, storageAccountName, storageAccountKey string) error {
+	fn, err := urlToFileName(url)
+	if err != nil {
+		return err
+	}
+
+	dl, err := getDownloader(
+		url,
+		storageAccountName,
+		storageAccountKey)
+	if err != nil {
+		return err
+	}
+
+	fp := filepath.Join(downloadDir, fn)
+	const mode = 0500 // we assume users download scripts to execute
+	if _, err := download.SaveTo(dl, fp, mode); err != nil {
+		return err
+	}
+
+	err = postProcessFile(fp)
```

Comment:
Why don't you pass in a func() instead of hard coding postProcessFile(). It _seems_ like a good extension point, but maybe it isn't necessary yet.